seamm_util/plotting.py

```diff
@@ -0,0 +1,554 @@
+# -*- coding: utf-8 -*-
+
+"""
+A package for handling plotly plots in SEAMM.
+
+A figure comprises one or more plots (or graphs) laid out in a grid.
+The datasets within a plot are represetned by traces, lines, scatter
+plots, pie chart elements, etc.
+
+This package consists of a Figure class which is a container, a Plot
+class that handles the actual plotting, and a Trace class for the
+individual datasets and their graphical representations. The figure
+class uses a grid similar to the Tk grid to control the layout of the
+plots.
+
+This package does not directly interact with plotly. Rather, it uses templates
+and Jinja to create the JSON, HTML or other representation of a plotly figure.
+The various containers in this package contain dictionaries of Jinja keys and
+the associated values. The keys needed are defined by the template used.
+"""
+
+from .dictionary import Dictionary
+import json
+
+
+class Figure(Dictionary):
+    """Holds one or more subplots and controls the layout.
+    """
+
+    def __init__(self, *args, jinja_env=None, template=None, **kwargs):
+        """Initialize a figure.
+
+        Keyword arguments:
+        """
+
+        super().__init__(*args, **kwargs)
+
+        self._jinja_env = jinja_env
+        self._template = template
+        self._plots = Dictionary(ordered=True)
+
+        self._grid = {'nrows': 0, 'ncolumns': 0, 'plots': {}, 'layout': {}}
+
+    @property
+    def template(self):
+        """The template to use for this figure."""
+        return self._template
+
+    @template.setter
+    def template(self, value):
+        self._template = value
+
+    def get(self, key, default=''):
+        """Return the value for key if key is in the dictionary, else
+        default. If default is not given, it defaults to an empty
+        string, so that this method never raises a KeyError.
+
+        Parameters
+        ----------
+        key : str
+            The key to fetch.
+        default : any value, optional
+            The value to return if the key is not in the data. Defaults to an
+            empty string.
+
+        Returns
+        -------
+        any
+            Returns the value associated with the key, or the default.
```

Comment:
Could documentation on this be clarified? It's confusing to say in the description that we are getting values for keys in the dictionary when we don't think of this object as a dictionary. I see that it inherits from our `Dictionary` class, so I assume this is why the description is written in this way, but I don't know that people using the class will know what dictionary we're talking about. I think this function is adding a short cut for getting dictionary values (can use `Plot.get` instead of `Plot.data.get` or `Plot.data[key_name]`? Alternatively, could we override the `__getitem__` function?

Reply:
I removed the method from Plot and Trace. I originally thought they would be used in Jinja templates, so override the Dictionary method, which had None as the default, with versions that had an empty string as default. Turns out these methods are not useful and caused confusion, so I removed them.

You should not use Plot.data.xxxx as that is an implementation detail, but should only use the plot methods. Figure, Trace and Axis are all dictionaries of values that are directly sent to Jinja, and need to be dictionaries because the keys are defined in the templates and hence aren't known in advance. Plot uses the dictionary to store the traces. which is perhaps not a great idea since it is different....